sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/datagen/DataGeneratorRowFn.java

```diff
@@ -0,0 +1,189 @@
+    this.schema = schema;
+    this.properties = properties;
+    this.primaryTimestampField = primaryTimestampField;
+  }
+
+  @Setup
+  public void setup() {
+    this.random = new Random();
+    this.fieldGenerators = new HashMap<>();
+
+    for (Schema.Field field : schema.getFields()) {
+      fieldGenerators.put(field.getName(), createGeneratorForField(field));
+    }
+  }
+
+  @ProcessElement
+  public void processElement(
+      @Element Long index, @Timestamp Instant timestamp, OutputReceiver<Row> out) {
+    Row.Builder rowBuilder = Row.withSchema(schema);
+    for (Schema.Field field : schema.getFields()) {
+      Object value;
+      if (field.getName().equals(this.primaryTimestampField)) {
+        value = timestamp.toDateTime();
+      } else {
+        FieldGenerator generator = fieldGenerators.get(field.getName());
+        if (generator == null) {
+          throw new IllegalStateException("No generator found for field: " + field.getName());
+        }
+        value = generator.generate(index);
+      }
+      rowBuilder.addValue(value);
+    }
+    out.output(rowBuilder.build());
+  }
+
+  @FunctionalInterface
+  private interface FieldGenerator extends Serializable {
+    @Nullable
+    Object generate(long index);
+  }
+
+  private FieldGenerator createGeneratorForField(Schema.Field field) {
+    String fieldName = field.getName();
+    FieldGenerator valueGenerator = createValueGeneratorForField(field);
+    double nullRate = properties.path("fields." + fieldName + ".null-rate").asDouble(0.0);
+
+    if (nullRate > 0) {
+      return (index) ->
+          Objects.requireNonNull(random).nextDouble() < nullRate
+              ? null
+              : valueGenerator.generate(index);
+    }
+    return valueGenerator;
+  }
+
+  private FieldGenerator createValueGeneratorForField(Schema.Field field) {
+    String fieldName = field.getName();
+    String kind = properties.path("fields." + fieldName + ".kind").asText("random");
+
+    if ("sequence".equalsIgnoreCase(kind)) {
+      JsonNode startNode = properties.path("fields." + fieldName + ".start");
+      JsonNode endNode = properties.path("fields." + fieldName + ".end");
+
+      if (startNode.isMissingNode() && endNode.isMissingNode()) {
+        return (index) -> index; // Simple, non-cycling sequence
+      }
+
+      if (startNode.isMissingNode() || endNode.isMissingNode()) {
+        throw new IllegalArgumentException(
+            "For a cycling sequence generator, both 'start' and 'end' must be specified.");
+      }
+
+      long start = startNode.asLong();
+      long end = endNode.asLong();
+
+      if (start > end) {
+        throw new IllegalArgumentException(
+            String.format(
+                "For sequence generator, 'start' (%d) cannot be greater than 'end' (%d).",
+                start, end));
+      }
+      long cycleLength = end - start + 1;
+      return (index) -> start + (index % cycleLength);
```

Comment:
Does this mean `sequence` only supports INT64 schema field type? If so, can we validate this ahead. of time?

Reply:
(I think we should be able to support other types as well, but doesn't have to be in this PR)